### ISSUE TITLE
add feature gate to permit measuring memory usage of dyn upcasting

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#112355]
@WaffleLapkin is preparing a PR to permit measuring memory usage of dyn upcasting to help gather data for concerns raised in https://github.com/rust-lang/rust/issues/65991. The lang team is generally in favor of gathering data but we don't want to wait infinitely long. Opening this issue to track this work.

## Instructions on how to measure trait upcasting effects on your codebase

You need a recent nightly compiler — 2023-06-14 or newer.

If you are using `cargo`, run `cargo clean && RUSTFLAGS="-Zprint-vtable-sizes" cargo check`. Note that `cargo clean` is required — without it caching will silence the output. If you are using `rustc` directly pass `-Zprint-vtable-sizes` to it.

The output will contain lines like these:
```
print-vtable-sizes { "crate_name": "num_traits", "trait_name": "RefNum", "entries": "23", "entries_ignoring_upcasting": "13", "entries_for_upcasting": "10", "upcasting_cost_percent": "76.92307692307693" }
print-vtable-sizes { "crate_name": "num_traits", "trait_name": "NumAssignOps", "entries": "12", "entries_ignoring_upcasting": "8", "entries_for_upcasting": "4", "upcasting_cost_percent": "50" }
print-vtable-sizes { "crate_name": "num_traits", "trait_name": "NumOps", "entries": "12", "entries_ignoring_upcasting": "8", "entries_for_upcasting": "4", "upcasting_cost_percent": "50" }
print-vtable-sizes { "crate_name": "num_traits", "trait_name": "cast::ToPrimitive", "entries": "17", "entries_ignoring_upcasting": "17", "entries_for_upcasting": "0", "upcasting_cost_percent": "0" }
print-vtable-sizes { "crate_name": "num_traits", "trait_name": "ops::inv::Inv", "entries": "4", "entries_ignoring_upcasting": "4", "entries_for_upcasting": "0", "upcasting_cost_percent": "0" }
print-vtable-sizes { "crate_name": "num_traits", "trait_name": "ops::mul_add::MulAdd", "entries": "4", "entries_ignoring_upcasting": "4", "entries_for_upcasting": "0", "upcasting_cost_percent": "0" }
print-vtable-sizes { "crate_name": "num_traits", "trait_name": "ops::mul_add::MulAddAssign", "entries": "4", "entries_ignoring_upcasting": "4", "entries_for_upcasting": "0", "upcasting_cost_percent": "0" }
print-vtable-sizes { "crate_name": "num_traits", "trait_name": "pow::Pow", "entries": "4", "entries_ignoring_upcasting": "4", "entries_for_upcasting": "0", "upcasting_cost_percent": "0" }
```

`print-vtable-sizes` followed by a json describing vtable sizes. Json will contain the following fields:
- `crate_name` — name of the crate which defines the traits
- `trait_name` — path to the trait
- `entries` — number of entries in a vtable with the current algorithm (i.e. with upcasting)
- `entries_ignoring_upcasting` — number of entries in a vtable, as-if we did not have trait upcasting
- `entries_for_upcasting` — number of entries in a vtable needed solely for upcasting (i.e. `entries - entries_ignoring_upcasting`).
- `upcasting_cost_percent` — cost of having upcasting in % relative to the number of entries without upcasting (i.e. `entries_for_upcasting / entries_ignoring_upcasting * 100%`).

Lines are sorted by `upcasting_cost_percent`, so that the biggest % change is first.

You can collect and analyze these stats however your want.

Some important notes:
1. This includes private traits
2. This includes traits that are not meant to be used as `dyn` and/or vtables of which are never actually instantiated
3. For traits with generics/associated types this can produce slightly misleading stats (only making the change bigger than it actually is) 
4. This does not include traits which are not object safe

**Comment by nikomatsakis** [comment]
@WaffleLapkin can you leave a comment with status update?

**Comment by WaffleLapkin** [comment]
I have a WIP branch [here](https://github.com/WaffleLapkin/rust/tree/vtable_stats), it needs a little bit of cleanup (like gating the stat print under `-Z` option) but overall works. I planned to finish cleanup today, but couldn't because of {personal reasons I don't think are worth disclosing here}. I'll try to finish the cleanup asap, hopefully tomorrow :)
<sub>I'm also not currently to run the experiment myself, because my laptop is too weak and dev desktops don't have any isolation like `podman`, but hopefully this will be sorted out soon....</sub>

**Comment by WaffleLapkin** [comment]
I've opened a PR: #112400 (sadly, while cleaning it up I've found a few ICEs I've introduced, which I'm not immediately sure how to fix...)

**Comment by WaffleLapkin** [comment]
The implementation PR was merged and with the latest nightly you can try this out (I've added instructions to the issue description).

Note, however, that I'm still not able to run a big experiment on top-N crater, or anything like that.

**Comment by saethlin** [comment]
Hi, Waffle asked if I could help it out so I'm giving this experiment a shot. I maintain a crater-like tool that makes a number of different design decisions. I'll upload a tarball in a day or so I think with output from the latest version of all published crates per the instructions above.

**Comment by Urgau** [comment]
I just ran the option with some well known crates and created a [script](https://gist.github.com/Urgau/e17b0bea9eae60c615d5758ac187d243) to summarises those results.

I may edit this post with more results in the future; also fell free to use my script.

-----

# Summary for axum-all-features, axum, bevy

Upcasting cost: MEAN ± STDEV (MIN, MAX)

*note: globally means that even traits that are not impacted are taken into account*

## axum-all-features (with dependencies)

Total entries: 3233 (for 708 prints)
Total entries for upcasting: 62 (~1.9%)

Upcasting cost: **~24.7%** ± 16.8 (min: 7.69, max: 76.9)
Upcasting cost (globally): **~0.872%** ± 5.51 (min: 0.0, max: 76.9)

## axum (with dependencies)

Total entries: 1622 (for 303 prints)
Total entries for upcasting: 19 (~1.2%)

Upcasting cost: **~20.3%** ± 9.21 (min: 8.33, max: 40.0)
Upcasting cost (globally): **~0.938%** ± 4.68 (min: 0.0, max: 40.0)

## bevy (with dependencies)

Total entries: 5165 (for 709 prints)
Total entries for upcasting: 155 (~3.0%)

Upcasting cost: **~24.0%** ± 17.5 (min: 1.42, max: 80.0)
Upcasting cost (globally): **~2.44%** ± 9.13 (min: 0.0, max: 80.0)

**Comment by saethlin** [comment]
The file is too big for GitHub, so here's a Google Drive link: https://drive.google.com/file/d/1nURLpnrRODv8Essd268nhhqM_APFwrrX/view

`tar xf raw.tar.xz` should extract about 120,000 directories with the name of each crate, and in each should be a file with the version of the crate that was built. The total extracted size is about 2.4 GB.

**Comment by LPGhatguy** [comment]
Howdy! We're working on a medium-sized commercial game in Rust. I ran the test on our codebase using `rustc 1.71.0 (8ede3aae2 2023-07-12)` and sorted the results by `upcasting_cost_percent`:

```bash
jq -s -c 'sort_by(.upcasting_cost_percent|tonumber) | reverse[]' output.json > sorted.json
```
Here's our results: https://gist.github.com/LPGhatguy/1a9624f0a30711f301460a777696bd5f

The biggest concern I would have is the impact on `wgpu_core` with ~80% upcasting cost. I haven't been following this feature or issue, so I'm not sure what impact that has on performance.

**Comment by nbdd0121** [comment]
I am not sure that this really captures the major cost of trait upcasting?

For example, if I have
```rust
trait A {
    fn foo(&self);
}

trait B {
    fn bar(&self);
}

trait C: A + B {}
```
and just uses `dyn C`, then without upcasting, vtable of `C` is 5 entries, and there's no need for vtable of `A` and `B`. With upcasting, the cost of `C` - 6 entries, and then vtable of `B` - 3 entries, which is 9 entries in total. The cost here is 80%, not 20% as the flag produces. The flag is underestimating the cost here.

It probably makes more sense to add a feature flag that disable trait upcasting computation at all and then just compare the number of vtables generated in the object files?

**Comment by tmandry** [comment]
> It probably makes more sense to add a feature flag that disable trait upcasting computation at all and then just compare the number of vtables generated in the object files?

I completely agree – the best way to measure would be to disable the feature and measure the end-to-end cost in the binary.

**Comment by nbdd0121** [comment]
Here's a few projects that I evaluated:

https://github.com/tock/tock (embedded):
```
  text    data     bss     dec     hex filename
160604      60   19888  180552   2c148 upcast.elf
  text    data     bss     dec     hex filename
160380      60   19888  180328   2c068 noupcast.elf
```
224 bytes increase in .text, 0.14% overhead. No .data overhead.

I evaluated a few smaller firmware/embedded projects, and I don't see size difference at all, since there are no `dyn` usage or `dyn` usage does not involve traits with multiple supertraits.

A personal WIP driver project:
```
   text	data     bss     dec     hex filename
 814169   12804   	0  826973   c9e5d upcast
   text	data     bss     dec     hex filename
 813233   12804   	0  826037   c9ab5 noupcast
```
936 bytes increase in .text, 0.12% overhead. 0.11% overall overhead.

https://github.com/nbdd0121/r2vm (hosted, size doesn't really matter, investigate out of curiosity):
```
   text    data     bss     dec     hex filename
3445057  238856    1584 3685497  383c79 upcast
   text    data     bss     dec     hex filename
3440257  236456    1584 3678297  382059 noupcast
```
4800 bytes increase in .text, 0.14% overhead. .data 2400 bytes, 1.01% overhead. 0.2% overall overhead for text + data combined.

---

With the data available, I think my conclusion is that the size overhead is <0.2% in all cases where I think size really matters a lot, so IMO the overhead is acceptable (when weighting against the language/compiler complexity to make the upcasting feature opt-in rather than always enabled).

My only concern now is about libcore/liballoc accidentally introduce upcastable traits and force the overhead to everyone, but that concern was already mitigated by the multiple_supertrait_upcastable lint that's enabled in libcore/liballoc.

I would encourage other people interested in binary size to give it a shot with the new flag and re-run the evaluation using the end-to-end approach and report numbers. @Urgau @LPGhatguy who posted some evaluation here and @oxalica who raised the initial vtable size concern.
